src/rtt.cpp

```diff
@@ -0,0 +1,590 @@
+            RETURN_ERROR_ON_FAIL(dll_function.rtt_is_control_block_found(&controlBlockFound), RTTCouldNotCallFunction);
+            std::chrono::high_resolution_clock::time_point attemptedStartupTime = std::chrono::high_resolution_clock::now();
+
+            if (controlBlockFound) {
+                break;
+            }
+
+            if (std::chrono::duration_cast<std::chrono::seconds>(attemptedStartupTime - rttStartTime).count() > 10) {
+                cleanup();
+                return RTTCouldNotFindControlBlock;
+            }
+        }
+
+        uint32_t downChannelNumber;
+        uint32_t upChannelNumber;
+
+        RETURN_ERROR_ON_FAIL(dll_function.rtt_read_channel_count(&downChannelNumber, &upChannelNumber), RTTCouldNotGetChannelInformation);
+
+        for(uint32_t i = 0; i < downChannelNumber; ++i)
+        {
+            char channelName[32];
+            uint32_t channelSize;
+            RETURN_ERROR_ON_FAIL(dll_function.rtt_read_channel_info(i, DOWN_DIRECTION, channelName, 32, &channelSize), RTTCouldNotGetChannelInformation);
+
+            std::string name(channelName);
+            baton->downChannelInfo.push_back(new ChannelInfo(i, name, channelSize));
+        }
+
+        for(uint32_t i = 0; i < upChannelNumber; ++i)
+        {
+            char channelName[32];
+            uint32_t channelSize;
+            RETURN_ERROR_ON_FAIL(dll_function.rtt_read_channel_info(i, UP_DIRECTION, channelName, 32, &channelSize), RTTCouldNotGetChannelInformation);
+
+            std::string name(channelName);
+            baton->upChannelInfo.push_back(new ChannelInfo(i, name, channelSize));
+        }
+
+        return RTTSuccess;
+    };
+
+    rtt_return_function_t r = [&] (RTTBaton *b) -> returnType {
+        auto baton = static_cast<RTTStartBaton*>(b);
+
+        returnType vector;
+
+        v8::Local<v8::Array> downChannelInfo = Nan::New<v8::Array>();
+        int i = 0;
+        for (auto element : baton->downChannelInfo)
+        {
+            Nan::Set(downChannelInfo, Convert::toJsNumber(i), element->ToJs());
+            i++;
+        }
+
+        vector.push_back(downChannelInfo);
+
+        v8::Local<v8::Array> upChannelInfo = Nan::New<v8::Array>();
+        i = 0;
+        for (auto element : baton->upChannelInfo)
+        {
+            Nan::Set(upChannelInfo, Convert::toJsNumber(i), element->ToJs());
+            i++;
+        }
+
+        vector.push_back(upChannelInfo);
+
+        return vector;
+    };
+
+    CallFunction(info, p, e, r);
+}
+
+void RTT::cleanup()
+{
+    if (!libraryLoaded)
+    {
+        return;
+    }
+
+    bool started;
+
+    dll_function.is_rtt_started(&started);
+
+    if (started)
+    {
+        dll_function.rtt_stop();
+    }
+
+    bool connected;
+
+    dll_function.is_connected_to_device(&connected);
+
+    if (connected)
+    {
+        dll_function.disconnect_from_emu();
+    }
+
+    dll_function.close_dll();
+
+    releasenRFjprog();
+
+    libraryLoaded = false;
```

Comment:
There might be an edge case / race condition here. Are there any instances where cleanup might occur twice in quick succession (in a shorter time than it takes for the DLL stuff to unload)?

Reply:
You are right in that race conditions may occur. The current implementation expects that the user is nice and never calls a function before the previous callback was completed. Will consider a fix for this.

Reply:
Have added an execution mutex that locks with a timeout of 10 seconds on all function calls. All race conditions should be handled by that. Two tests were written to test this, and one of them confirms that the guards work by first calling start, and then call read three times in a row. With previous implementation, the reads would return with an error that rtt was not started. Currently they return without error. 

Attempting to hit the 10 second timeout for the mutex proved harder. Even with 1M calls to read in a row as fast as possible, you would never get an error. But I have implemented the same functionality for nRFjprog, and when I attempt to program multiple times as fast as possible, the timeout triggers. This can be seen in `__tests__/racecondition.test.js`